### ISSUE TITLE
feat: コンテキスト長を最大65,536トークンまで設定可能にする

### ISSUE THREAD

**Issue by Kewton** [Kewton/photon-mlx#55]
## 背景・課題

現在 `max_position_embeddings: 2048` がハードコードされており、設定ファイルで 2048 以上の値を指定してもRoPEテーブルが 2048 までしか事前計算されないため、長いコンテキストを扱えない。

### 現在の制約

```yaml
# configs/photon_tiny.yaml / photon_small.yaml / photon_600m_paper.yaml
model:
  max_position_embeddings: 2048  # ← 全プロファイルで 2048（ハードリミット）
  rope_theta: 1000000.0

training:
  context_length: 2048           # ← データパックの上限（全プロファイルで 2048）
```

```python
# photon_mlx/model.py:155-157 — __init__ で固定長RoPEテーブルを生成
self._rope_cos, self._rope_sin = precompute_rope(
    m.head_dim, m.max_position_embeddings, m.rope_theta  # 2048まで
)
# photon_mlx/model.py:158-160 — local RoPE も max_local 長で precompute
# （両方のテーブル更新を漏らさないこと）
```

`photon_mlx/blocks.py:16-25` の `precompute_rope()` は vanilla RoPE のみで、NTK-aware scaling や position interpolation は非対応。同名関数 `_precompute_rope` が `torch_ref/model.py:34` にも独立実装されている。

## 前提条件（クリア済み）

#58「PHOTONへの入力を質問文＋証拠パックに変更」は **PR #69 / commit 720953d で既にマージ済み**。現在 PHOTON は `question + "\n\n" + evidence_text` を受け取っており、`evidence_pack.max_tokens = 16,000` に設定された環境では 2048 の位置エンベディング上限が実質的ボトルネックになっている。

```python
# baseline_reporag/photon_pipeline.py:484 — 現状（#58後）
photon_input_text = question + "\n\n" + evidence_text_for_photon
```

本Issue #55 の解消により、この上限が外れ、長コンテキスト入力が有効化される。

## スコープ境界（重要）

本Issue #55 は **PhotonModel (MLX, `photon_mlx/`) のみ** を対象とする。torch_ref (`torch_ref/`) は PyTorch リファレンス実装で、`MinimalLM.__init__` が `register_buffer('causal_mask', (max_position_embeddings, max_position_embeddings))` で **dense な causal mask を事前確保**する（`torch_ref/model.py:177-183`）。65,536 を指定すると `65536² × 4B ≈ 17.2 GB` の単一テンソルが確保され、256GB 統合メモリでも OOM 相当になる。

したがって本Issueのスコープ境界は以下:

- **スコープ内**: `photon_mlx/` (MLX側)、`configs/photon_long_context.yaml`（新規）、`baseline_reporag/photon_pipeline.py` の ModelConfig 再構築更新、関連 bench/docs
- **スコープ外**: `torch_ref/` は従来通り `max_position_embeddings=128`（tiny_config）で回帰する。`torch_ref/model.py:34` の `_precompute_rope` は **シグネチャだけ** `scaling` / `scale_factor` を受け付けるように拡張し、デフォルト `"none"` で既存挙動を完全維持する（PhotonModel との API 整合のためだけの最小変更）

## ⚠️ 重要：推論時の拡張と学習の関係

PHOTON モデルは `context_length: 2048` で学習されているため、`max_position_embeddings` を 65,536 に拡張しても**学習時に見たことのない位置**で推論することになる。以下の2通りの対応が必要：

### 対応A：position interpolation / NTK-aware scaling（v1 — 本Issueのスコープ）

- `precompute_rope()` に NTK-aware scaling と linear interpolation のオプションを追加
- 推論時に `rope_theta` を引き上げる（例: `rope_theta: 10000000.0`、YAML互換な数値リテラルで記述すること）
- 完全な再学習は不要だが、精度劣化リスクあり
- `ModelConfig` に `rope_scaling` / `rope_scale_factor` フィールドを追加

### 対応B：長いコンテキストで再学習（v2 — 別Issueに切り出し）

- `context_length` を拡張した値（目標値・curriculum 要否・学習時間見積は v2 Issue で決定）で PHOTON を再学習
- データは既に `pack_sequences()` で対応可能
- 最も安全だが学習コストが高い（現在の学習時間は約3時間@1000step @ 2048）
- v2 は「Issue: 長コンテキスト再学習」として別途切り出す。本Issue #55 は v1 に限定。

## メモリ可用性の見積り（要実測）

PHOTONは階層的アテンション（トップレベルは T/16 位置で二次）のため、標準Transformerの1/16のメモリで長コンテキストを処理できるが、具体的な実測値は未検証。

**算出前提**（暫定見積り）:
- `RoPE table = 2 × max_len × head_dim/2 × 4B (fp32)`（例: 65536 × 64/2 × 4B × 2 ≈ 16MB）
- `top-level attention score = layers × heads × (T/16)² × 2B (fp16)`
- **KV cache（PR #71）**: `layers × n_kv_heads × (T/16) × head_dim × 4B (K+V, fp16)` に加え `cached_top_out: hidden_size × (T/16) × 2B` が `generate()` 経路でデフォルト有効（`photon_small` config で 32768 入力時、約 50-100MB オーダー）
- batch=1、activation は除外
- 推論時のみ

| コンテキスト長 | 推論メモリ（KV cache 込み, 見積り） | 備考 |
|--------------|-------------|------|
| 2,048（現在） | ~0.71 GB | 既存実測相当 |
| 8,192 | ~0.75 GB | 見積り |
| 32,768 | ~0.88–0.95 GB | KV cache の分 +50MB 想定 |
| **65,536** | **~1.10–1.30 GB** | KV cache の分 +100MB 想定 |

実測は `bench/` ハーネスで実装後に確認する（詳細は「実装内容 / 7. bench 検証」）。256GB 統合メモリ環境では、上記見積りが2桁誤っても余裕がある想定。KV cache による RAM 圧迫が問題になる場合は `use_kv_cache=False` へのフォールバック手順を `docs/troubleshooting.md` に記載する。

### 1. 設定ファイルの扱い

**既存3プロファイル**（`photon_tiny.yaml`, `photon_small.yaml`, `photon_600m_paper.yaml`）は `max_position_embeddings: 2048`、`training.context_length: 2048` を**維持**する（既存チェックポイントとの互換性のため）。**ModelConfig / TrainingConfig の default 値も変更しない**（torch_ref/tests/test_model.py の default snapshot テストの破壊を避けるため）。

**新規プロファイル** `configs/photon_long_context.yaml` を追加:

```yaml
model:
  max_position_embeddings: 65536
  rope_theta: 10000000.0           # YAML互換な数値リテラル（アンダースコア不可）
  rope_scaling: ntk                # v1 allowed values: "none", "ntk"
  rope_scale_factor: 32.0          # 例: 2048 → 65536 は 32倍
training:
  context_length: 32768            # v1 検証用。本番は 65536 も可
```

### 2. ModelConfig の拡張とバリデーション（`torch_ref/config.py`）

`ModelConfig` に次のフィールドを追加（**default値は下位互換を維持**）：

```python
rope_scaling: str = "none"  # v1 choices: {"none", "ntk"}
rope_scale_factor: float = 1.0
```

以下のバリデーションを追加:

- `context_length % prod(chunk_sizes) == 0`（デフォルト `chunk_sizes=[4,4]` なので 16 の倍数であること）
- `max_position_embeddings >= context_length`
- `rope_scaling in {"none", "ntk"}` かつ `rope_scale_factor >= 1.0`（v1 では linear は非採用）

### 3. データパイプラインの対応（`photon_mlx/data.py`）

`pack_sequences()` は既に `context_length` パラメータを受け取るため変更不要。ただし次の注意点を実装コメントで明記：

> **注**: `pack_sequences()` は残余（`< context_length` のバッファ）を捨てる（data.py:37 `Drop remainder shorter than context_length`）。`context_length` を 2048 → 32768 に引き上げると小文書コーパスでは packed sample が大幅に減る可能性がある。v2（再学習）の前提として corpus size と実効 sample 数の事前計測が必要。

### 4. inference.py の上限チェック — 対応不要（既に実装済み）

`photon_mlx/inference.py:240` の `max_len = self.cfg.model.max_position_embeddings` は既に config 参照になっており、hardcoded な 2048 は存在しない。**本Issueでの追加対応は不要**。ただし、新しい `max_position_embeddings` 値が end-to-end で wired されることを回帰テストで担保する。

### 5. 位置補間の実装（`photon_mlx/blocks.py` + `torch_ref/model.py` 同期更新）

**MLX 側**: `photon_mlx/blocks.py::precompute_rope()` のシグネチャを以下に拡張:

```python
def precompute_rope(
    dim: int,
    max_len: int,
    theta: float = 1_000_000.0,
    *,
    scaling: str = "none",           # v1: {"none", "ntk"}
    scale_factor: float = 1.0,
) -> tuple[mx.array, mx.array]:
    """
    - scaling="none": vanilla RoPE（既存挙動、下位互換）
    - scaling="ntk":  theta を theta * scale_factor ** (dim / (dim - 2)) に置き換え（NTK-aware Interpolated RoPE）
    Note: scaling="ntk" + scale_factor=1.0 is mathematically equivalent to scaling="none".
    """
```

**v1 採用手法**: `ntk` のみ採用。`scaling="none"` をデフォルト引数として既存挙動を維持し、既存テストを破壊しない。linear は v1 非採用（設計レビュー DR1-005 の判断）。NTK が実測で不十分な場合は v2（再学習 or YaRN 等）に倒す。

**PyTorch リファレンス側**: `torch_ref/model.py:34::_precompute_rope` も同じ `scaling` / `scale_factor` kwargs を受け付けるように拡張する。**ただし** torch_ref 側のデフォルトは `scaling="none"` 固定とし、`MinimalLM` 経路（max_position_embeddings=128 の tiny_config）は挙動変更しない。本Issueで torch_ref を 65536 で動かすことは受入条件に含めない。

**PhotonModel 側**: `photon_mlx/model.py:155-160` の `precompute_rope()` 呼び出しを、`ModelConfig.rope_scaling` / `rope_scale_factor` / `context_length` を wire するように更新する。アクセスは `getattr(m, "rope_scaling", "none")` / `getattr(m, "rope_scale_factor", 1.0)` で行い、MagicMock 経由の cfg を使う既存テスト (`baseline_reporag/tests/test_photon_pipeline.py` など) が破壊されないようにする。local RoPE (`158-160`) にも同様の scaling を適用する。

### 6. baseline_reporag 経由の config 伝播（`baseline_reporag/photon_pipeline.py`）

`_build_photon_deps`（`photon_pipeline.py:169-176`）での `ModelConfig` 再構築には現在 `max_position_embeddings` / `rope_theta` が wire されていない（= default の 2048 / 1e6 に巻き戻る）。以下のフィールドを明示的に引き継ぐよう更新:

- `max_position_embeddings`, `rope_theta`, `rope_scaling`, `rope_scale_factor`, `head_dim`

baseline YAML 側がこれらを持たない場合は photon_cfg.model の値を fallback として使う。

### 7. bench 検証

長コンテキストの実測として以下を実施:

- `bench/kv_cache_speedup.py --config configs/photon_long_context.yaml --prompt-len 16384 --gen 8` を実行し、speedup 結果を `reports/issue-55-long-context.md` として保存
- `bench/issue61_prune_batch.py --config configs/photon_long_context.yaml --max-len 32768` を一度走らせ、prune batch 経路が 32768 で破綻しないことを確認

### 8. ドキュメント更新

- `docs/tutorial.md`: 「長コンテキスト推論を有効化する」セクションを新設。`configs/photon_long_context.yaml` の使い方、`evidence_pack.max_tokens` の推奨値、`use_kv_cache=False` へのフォールバック手順を記載
- `docs/troubleshooting.md`: 「RAM 不足時」の節を追加し、KV cache オフ化の手順を記載

### コード

- [ ] `configs/photon_long_context.yaml` で `max_position_embeddings: 65536` / `rope_scaling: ntk` / `rope_scale_factor: 32.0` を指定して `PhotonModel.__init__` が例外なく完了する
- [ ] `context_length: 32768` でデータパックが正常に実行できる（`pack_sequences(..., context_length=32768)` が動作）
- [ ] `context_length` が `prod(chunk_sizes)` の倍数でない場合、`ModelConfig` バリデーションがわかりやすいエラーを出す
- [ ] `context_length > max_position_embeddings` の場合、`ModelConfig` バリデーションが起動時エラーを出す
- [ ] `rope_scaling` が `{"none", "ntk"}` 以外の場合、バリデーションエラー（v1 では linear は非採用）
- [ ] `rope_scale_factor < 1.0` の場合、バリデーションエラー

### 下位互換

- [ ] 既存 `ModelConfig` / `TrainingConfig` の default 値は変更しない（`max_position_embeddings: int = 2048`, `rope_theta: float = 1e6`, `context_length: int = 2048`）
- [ ] 新フィールド `rope_scaling: str = "none"`, `rope_scale_factor: float = 1.0` の default も下位互換を壊さない値
- [ ] 既存の 2048 で学習したチェックポイントを `photon_long_context.yaml` (max_position_embeddings=65536) の config で load して推論できる
- [ ] **different-max_position_embeddings 経路**の回帰テストを `photon_mlx/tests/test_training.py` に新規追加（例：tiny config で保存→long_context config で load→generate(512, 4) が成功）
- [ ] `baseline_reporag/tests/test_photon_pipeline.py` が **無変更**で全パスする（MagicMock 経由 cfg が新フィールド追加で壊れないこと）
- [ ] `torch_ref/tests/test_model.py` 全 11 テストが挙動変更なくパスする（default snapshot 維持）
- [ ] `precompute_rope()` の `scaling="none"` デフォルトで既存テスト（test_rope_* 等）が挙動変更なくパスする

### NTK 動作検証

- [ ] `rope_scaling="ntk"` / `rope_scale_factor>=2.0` を指定したとき、2048 で学習した checkpoint を load した `PhotonModel` で、長さ 8192 の入力に対する `generate()` が例外なく完了する
- [ ] YAML の `model.rope_scaling` / `model.rope_scale_factor` が `ModelConfig` および `precompute_rope()` に wire されていることを確認する統合テスト
- [ ] baseline_reporag 経由で `configs/photon_long_context.yaml` の設定値が `PhotonModel` に反映される統合テスト

### bench / docs

- [ ] `bench/kv_cache_speedup.py --config configs/photon_long_context.yaml --prompt-len 16384` が実行でき、結果を `reports/issue-55-long-context.md` に記録
- [ ] `bench/issue61_prune_batch.py --config configs/photon_long_context.yaml --max-len 32768` が破綻なく完了
- [ ] `docs/tutorial.md` / `docs/troubleshooting.md` に新 profile のエンドツーエンド使用例と RAM 節約手順が記載される

### 全体

- [ ] 既存テスト（`photon_mlx/tests/`, `torch_ref/tests/`, `baseline_reporag/tests/`）が全パス（現状 72+/72+）
- [ ] `ruff check .` 警告0件、`ruff format --check .` 差分なし

## Future Work（本Issueのスコープ外）

- **v2 再学習**: `context_length` を拡張した値での PHOTON 再学習（別Issueに切り出し）
- **evidence_pack の tokenize-based 再パッキング**: 現状 `baseline_reporag/generation/evidence_pack.py:57` は `char_budget = max_tokens * 4` の粗い文字ヒューリスティックで予算を決定。65k まで引き上げた際、日本語/コード比率の高いコーパスでは 20-30% の over-truncation が発生する可能性あり。v2 で tokenize 後の再パッキングに切り替える

## 関連Issue

- #54 KVキャッシュ実装（長コンテキストでの速度問題を合わせて解決）
- ~~#58 PHOTONへの入力を質問文＋証拠パックに変更~~ — **PR #69 で既にマージ済み（前提条件クリア）**
- #56 2パス検索（Pass 1 のチャンク数拡大に長コンテキストが有効）
- 新規（v2）: 長コンテキスト再学習（本Issue完了後に別Issueで切り出し）
